### PR TITLE
[fix][broker] Gracefully shutdown does not work with admin cli in standalone

### PR DESCRIPTION
**Pull request by JooHyukKim**
Fixes : #20617

### Motivation

Currently, clients' shutdown API does not behave consistently in sense that asynchronicity is not handled explicitly. So issue #20617 happens.

This PR will allow clients know that shutdown is actually triggered.

### Modifications

- Synchronize call to `POST /shutdown` on client side
- Asynchronize explicitly `pulsar().closeAsync()` invocation
- 
### Verifying this change

- [x] Make sure that the change passes the CI checks.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [x] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/JooHyukKim/pulsar/pull/17

### Note

Here is [link to the original PR](https://github.com/apache/pulsar/pull/20651), that ended up in messed up state wrt merge.

### CODE REVIEW

**Review by Technoboy- — CHANGES_REQUESTED** [review]
A simple way to fix this is to `wait` the future.
```
getAdmin().brokers().shutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic).join();
```

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/BrokersImpl.java [hunk 1]
```diff
@@ -210,12 +210,13 @@ public void failed(Throwable throwable) {
...
-    public CompletableFuture<Void> shutDownBrokerGracefully(int maxConcurrentUnloadPerSec,
-                                                            boolean forcedTerminateTopic) {
+    public void shutDownBrokerGracefully(
+            int maxConcurrentUnloadPerSec, boolean forcedTerminateTopic
+    ) {
         WebTarget path = adminBrokers.path("shutdown")
                 .queryParam("maxConcurrentUnloadPerSec", maxConcurrentUnloadPerSec)
                 .queryParam("forcedTerminateTopic", forcedTerminateTopic);
-        return asyncPostRequest(path, Entity.entity("", MediaType.APPLICATION_JSON));
+        asyncPostRequest(path, Entity.entity("", MediaType.APPLICATION_JSON)).join();
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/BrokersImpl.java:219]: Not here.
Fix #20617,  I think the simple way is to fix like 
```
getAdmin().brokers().shutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic).join();
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 2]
```diff
@@ -545,16 +545,27 @@ public void shutDownBrokerGracefully(
...
+                .runAsync(
+                        () -> {
+                            pulsar().closeAsync();
+                            LOG.info("Broker graceful shutdown successfully");
+                        })
+                .exceptionally(ex -> {
+                    LOG.error("Broker graceful shutdown failed", ex);
+                    return null;
+                });
+        asyncResponse.resume(Response.noContent().build());
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:568]:
```
private CompletableFuture<Void> doShutDownBrokerGracefullyAsync(int maxConcurrentUnloadPerSec, boolean forcedTerminateTopic) {
 pulsar().getBrokerService().unloadNamespaceBundlesGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic);
  return pulsar().closeAsync();
}
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 3]
```diff
@@ -545,16 +545,27 @@ public void shutDownBrokerGracefully(
             @ApiParam(name = "maxConcurrentUnloadPerSec",
                     value = "if the value absent(value=0) means no concurrent limitation.")
             @QueryParam("maxConcurrentUnloadPerSec") int maxConcurrentUnloadPerSec,
-            @QueryParam("forcedTerminateTopic") @DefaultValue("true") boolean forcedTerminateTopic
+            @QueryParam("forcedTerminateTopic") @DefaultValue("true") boolean forcedTerminateTopic,
+            @Suspended final AsyncResponse asyncResponse
     ) {
         validateSuperUserAccess();
-        doShutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic);
+        doShutDownBrokerGracefullyAsync(maxConcurrentUnloadPerSec, forcedTerminateTopic, asyncResponse);
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:552]: doShutDownBrokerGracefullyAsync().thenAccept(__ -> asyncResponse.resume(Response.noContent().build()));

Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java]: And need to add `exceptionally`

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdBrokers.java [hunk 4]
```diff
@@ -150,8 +150,8 @@ private class ShutDownBrokerGracefully extends CliCommand {
 
         @Override
         void run() throws Exception {
-            getAdmin().brokers().shutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic);
-            System.out.println("Successfully trigger broker shutdown gracefully");
+            getAdmin().brokers().shutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic).join();
+            System.out.println("Successfully shutdown broker gracefully");
```
Comment (Technoboy-) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdBrokers.java:154]: I find that we have many async calls in the cli, we'd better add a `sync` call the CmdBase:
```
protected long getReadTimeoutMs() {
        PulsarAdmin pulsarAdmin = getAdmin();
        if (pulsarAdmin instanceof PulsarAdminImpl) {
            return ((PulsarAdminImpl) pulsarAdmin).getClientConfigData().getReadTimeoutMs();
        }
        return 60000;
    }

    protected <T> T sync(Supplier<CompletableFuture<T>> executor) throws PulsarAdminException {
        try {
            return executor.get().get(getReadTimeoutMs(), TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new PulsarAdminException(e);
        } catch (TimeoutException e) {
            throw new PulsarAdminException.TimeoutException(e);
        } catch (ExecutionException e) {
            throw PulsarAdminException.wrap(getApiException(e.getCause()));
        } catch (Exception e) {
            throw PulsarAdminException.wrap(getApiException(e));
        }
    }
```

And change this sync:
```
sync(() -> getAdmin().brokers().shutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic));
            System.out.println("Successfully trigger broker shutdown gracefully");
```

What's your idea ?  @codelipenghui @poorbarcode @mattisonchao @coderzc

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java (hunk 3 above)
Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java]: It seems can't give a response after close, see:https://github.com/apache/pulsar/pull/14114#issuecomment-1029853606

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 5]
```diff
@@ -545,16 +545,26 @@ public void shutDownBrokerGracefully(
...
+            @Suspended final AsyncResponse asyncResponse
     ) {
-        validateSuperUserAccess();
-        doShutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic);
+        validateSuperUserAccessAsync()
+                .thenCompose(__ -> doShutDownBrokerGracefullyAsync(maxConcurrentUnloadPerSec, forcedTerminateTopic))
+                .thenAccept(__ -> {
+                    LOG.info("[{}] Successfully shutdown broker gracefully", clientAppId());
+                    asyncResponse.resume(Response.noContent().build());
+                })
```
Comment (labuladong) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:556]: `doShutDownBrokerGracefullyAsync` method will close the web service. After the web service is closed, can `asyncResponse.resume(Response.noContent().build());` work? I think we can resume the response first, then call `doShutDownBrokerGracefullyAsync`. What do you think?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 6]
```diff
@@ -545,16 +545,27 @@ public void shutDownBrokerGracefully(
             @ApiParam(name = "maxConcurrentUnloadPerSec",
                     value = "if the value absent(value=0) means no concurrent limitation.")
             @QueryParam("maxConcurrentUnloadPerSec") int maxConcurrentUnloadPerSec,
-            @QueryParam("forcedTerminateTopic") @DefaultValue("true") boolean forcedTerminateTopic
+            @QueryParam("forcedTerminateTopic") @DefaultValue("true") boolean forcedTerminateTopic,
+            @Suspended final AsyncResponse asyncResponse
     ) {
         validateSuperUserAccess();
-        doShutDownBrokerGracefully(maxConcurrentUnloadPerSec, forcedTerminateTopic);
+        CompletableFuture.runAsync(
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:552]: We can remove `CompletableFuture.runAsync`.
Just like :
```
doShutDownBrokerGracefullyAsync(maxConcurrentUnloadPerSec, forcedTerminateTopic)
                                .thenAccept(__ -> {
                                    LOG.info("[{}] Successfully shutdown broker gracefully", clientAppId());
                                    asyncResponse.resume(Response.noContent().build());
                                }))
                .exceptionally(ex -> {
                    LOG.error("[{}] Failed to shutdown broker gracefully", clientAppId(), ex);
                   resumeAsyncResponseExceptionally(asyncResponse, ex);
                    return null;
                });
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 7]
```diff
@@ -545,16 +545,27 @@ public void shutDownBrokerGracefully(
...
+                        () -> doShutDownBrokerGracefullyAsync(maxConcurrentUnloadPerSec, forcedTerminateTopic)
+                                .thenAccept(__ -> {
+                                    LOG.info("[{}] Successfully shutdown broker gracefully", clientAppId());
+                                    asyncResponse.resume(Response.noContent().build());
+                                }))
+                .exceptionally(ex -> {
+                    LOG.error("[{}] Failed to shutdown broker gracefully", clientAppId(), ex);
+                    return null;
+                });
+        asyncResponse.resume(Response.noContent().build());
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:562]: Remove this line

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java (hunk 5 above)
Comment (labuladong) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java]: I don't think my question is similar to https://github.com/apache/pulsar/pull/20709#discussion_r1252482006. I mean, if we run `doShutDownBrokerGracefullyAsync` first, then run `asyncResponse.resume(Response.noContent().build());`, is that OK? In `doShutDownBrokerGracefullyAsync` method, the admin web service will be shut down, can `asyncResponse.resume` succeed when the web service has shut down?

I'm not sure if my guess is accurate, but I remember encountering a similar situation before. I suggest run `asyncResponse.resume(Response.noContent().build());` first, then run `doShutDownBrokerGracefullyAsync`.

**Review by Technoboy- — APPROVED** [review]
LGTM
@mattisonchao @codelipenghui @poorbarcode

**Review by liangyepianzhou — APPROVED** [review]

**Review by poorbarcode — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdBase.java [hunk 8]
```diff
@@ -26,17 +27,28 @@
...
 public abstract class CmdBase {
     protected final JCommander jcommander;
     private final Supplier<PulsarAdmin> adminSupplier;
     private PulsarAdmin admin;
     private IUsageFormatter usageFormatter;
 
+    /**
+     * Default read timeout in milliseconds.
+     * Used if not found from configuration data in {@link #getReadTimeoutMs()}
+     */
```
Comment (Anonymitaet) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdBase.java:49]: @JooHyukKim this PR is labeled with `doc-required`, and explanations have been added to API code files. Does it mean that the doc has been already added?

Comment (Anonymitaet) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdBase.java:49]: OK

**Pull request merged**